SSA: fix ssa to cssa, and naive out of cssa

serpilliere opened the pull request:
This PR implements the common first step of "Translating Out of Static Single Assignment Form" and "Revisiting Out-of-SSA Translation for Correctness, Code Quality, and Efficiency" 

This patch fixes the current Miasm UNSSA algorithm, which is buggy.
 PR done with @commial. Big thanks to you man!

serpilliere commented:
I remove the WIP as it seems we have interesting results here :smile:

serpilliere commented:
WIP because I am tracking a possible bug...

serpilliere commented:
Ok. It was a missuse of `immutable_ids` and `ssa_variables_expr`
Now, both are correctly used :smile: 

I used this PR to change the IDA `graph_ir` plugin, which is now a bit more customable and complete!

commial commented:
For those interested, here's what the GraphIR customization looks like with this PR:
![ida](https://user-images.githubusercontent.com/4194483/50398384-6d6eee80-0777-11e9-8458-27a9a33047ed.png)

@ecx86, regarding #823 it seems that you've already played with Boissinot out-of-ssa. If you had time, it would be nice to hear your opinion on this implementation and/or if you have specific test cases that you want to try it on. To be more specific, this implementation also integrate the "phi-virtualisation" optimization indicated in the paper (to be able to have a pre-computed liveness in input).

commial commented on example/ida/graph_ir.py:
```diff
@@ -2,23 +2,81 @@
+
+Options:
+<Simplify code:{rCodeSimplify}>
+<Subcall dont change stack:{rDontModStack}>
```
```suggestion
<Subcalls don't change stack:{rDontModStack}>
```

serpilliere commented:
@commial : I fixed the liveness analysis: the "phi" liveness was badly computed.
But there is a remaining bug in our implementation of the unssa algorithm: We have a minimalist example in `graph_8.dot` of `unssa.py` example.

My analysis: when we remove the `phi` of the `b` variable, we do this for each parents; here, lbl0/lbl2/lbl3. We generate the 'parallel copy' at the end of each of this parents; but when we generate the copy for block `lbl2`, we affect `b.0` which will interfere with the `b.0` at  `lbl3`

The liveness analysis seems good to me now. But I think  we have coalesced to quickly the `phi` predecessor variable (Maybe it's not clear, but it's not clear in my head neither...)
![final_8](https://user-images.githubusercontent.com/9433661/50528963-cdf08980-0af1-11e9-844c-6c5b4fcdeb10.png)

serpilliere commented:
Here is a graph (manually crafted...) with the parallel copy added:
![ref_8_vars](https://user-images.githubusercontent.com/9433661/50529976-8e2da000-0af9-11e9-8adf-cd0dc9df16cb.png)
And the same just afeter the trivial phi coalesced:
![ref_8_coalesce](https://user-images.githubusercontent.com/9433661/50529979-9c7bbc00-0af9-11e9-890f-97762815511d.png)
Here, it's clear we can coalesce `b` with `b.0`.
It seems we can coalesce `b` with `b.3` as well, but I don't see the possibility of coalescing `b` with `b.1`

serpilliere commented:
Ok, first bug: 
```python
                if merge_a_list[-1] < merge_b_list[-1]:
```
We are doing a 'string' comparison instead of comparing the `live_index` (lol)
But the result is not good anyway. We have to dig into the `live_index`...
I am pushing the modification to compare the `live_index`

serpilliere commented:
I think I have the bug: When we should create temporary variables, during the phi analysis, we should create pre `phi` variables (Here, `var0`, `var1`, `var2` for `b` phi, and  `var5`, `var6`, `var7` for `a` phi).
But we didn't create them, we only created post `phi`. So pre phi are not taken into account during the interference test: we merge `var0`, `var1`, `var2` with `var3` without putting `var0`, `var1`, `var2` in the new merge set.
The resulting interference computation is incorrect.

serpilliere commented:
The live check against the `varX` (which we don't need to create) is defined by the sentence:
```
Also,
to check the intersection with other variables, it is sufficient
to check the intersection at the parallel copies locations
```

But when we test the liveness of `b.0` and `var1`, we only test the liveness at the *post-phi* definition of `var1`, which does not comply with the paper: we have to check against *each* parallel copies, even the "pre-phi" locations.
So for liveness test, we have to check if one variable is a post phi, if so, check against every pre phi locations as well. Today, we check against only for one phi predecessor. this may be wrong

serpilliere commented:
I added the liveness test against pre phi definition, and it seems to work :tada: 
I will re analyse the code to be sure, and push the PR
Here is the result:
![final_8](https://user-images.githubusercontent.com/9433661/50547896-9a257900-0c43-11e9-9808-abb60d73354e.png)
I may have missed a case as for me `a.0` and `a.2` should merge, but, erf, the result seems good :smile:

serpilliere commented:
Here is the last fix: I enable the expression propagation in the unssa regression tests, which generate interesting cases for it :smile:

commial commented:
Hey,

I agree with your different analyses:
* String comparison for merge_list... I guess it was working because variables are created in the dominance order, so `a.X` index were quite reflecting `live_index`...
* I do agree with the fact that we have to check for interference for pre-phi, and doing it in a way that all predecessors are checked before taking any decision
* Regrading your last example, I also agree with you. Live range of `a.2` is only in `lbl3`, so they should be able to interfere. It might help to have a look in the "interfere" test to check why they are considered as interfering

I should also have a look at the whole resulting code, because reviewing the different diffs make me loose the big picture.
I also wonder what happen if a given ssa var is used in two different phi, like:
```
                      a.1 = X
                     _______
                     /     \
| a.2 = phi(a.1, ...) |    | a.3 = phi(a.1, ...) |
```

I can't test the code for now, will do the next year.

serpilliere commented:
No problem @commial. I whish I pushed the different commits before merging them.
Maybe I can get back the temp commits.

I forgot to send the pic of the final last graph: I seems to be good.
![final_8](https://user-images.githubusercontent.com/9433661/50562585-4e430480-0d15-11e9-80f3-82e7b1212ec2.png)

serpilliere commented:
I have fixed an inversion which I did during some tests. This broke classic cases. I added another regression test to ensure this.
Note: In fact the regression tests are not "tested" here: I didn't find yet an interesting way of verifying the results.

commial commented on miasm2/analysis/ssa.py:
```diff
@@ -575,120 +574,505 @@ def _convert_phi(self):
-    @assignblk_dct: dictionary representing the AssignBlock
+def get_var_assignment_src(ircfg, node, variables):
     """
+    Return a subset of @variables which are written by the irblock at @node
```
It actually ensure and returns the only one variable of `variables` written in the assignblk (ie, the subset is always zero or one var)
